six/modules/python/cphd/source/cphd.i

```diff
@@ -143,6 +144,289 @@ using six::Vector3;
+            channel_data = channel_data.reshape(num_vectors, pvp_size)
+
+            pvp_offset = 0
+            for name, (size, offset, fmt) in sorted_pvp_info.items():
+                suffix_to_format = PVPBlock._split_multi_format(fmt)
+
+                # Number of values in this parameter, not necessarily equal to
+                # _size (e.g. 'CF16' uses 2 8-byte words for 1 value)
+                # Store this in case suffix_to_format changes below
+                num_values = len(suffix_to_format)
+
+                # If a PVP has multiple parameters but all have the same
+                # format, collapse them into a single PVP. This is intended
+                # to simplify things like 'txPos', where it's more convenient
+                # to have an (N,3)-shape array than three length-N arrays
+                # for 'txPosX', 'txPosY', and 'txPosZ' separately.
+                _size = 1
+                formats = list(suffix_to_format.values())
+                if PVPBlock._common_format(formats):
+                    suffix_to_format = {'': formats[0]}
+                    _size = size
+
+                for suffix, _fmt in suffix_to_format.items():
+                    _name = name + suffix
+                    dtype = PVPBlock._format_to_dtype(_fmt)
+                    length = dtype.itemsize * num_values
+
+                    # For every vector (first index of channel_data), slice
+                    # the range with values for this parameter (second index)
+                    vals = channel_data[
+                        :, pvp_offset:(pvp_offset + length)
+                    ].copy().view(dtype)
+                    channel_pvp[_name] = vals.reshape(
+                        num_vectors, num_values
+                    ).squeeze()
+                    pvp_offset += 8 * _size  # Move to next 8-byte word
+
+            pvp_data.append(channel_pvp)
+
+        return pvp_data
+
+    @staticmethod
+    def from_list_of_dicts(pvp_data, metadata):
+        """Initialize and populate a PVPBlock from a list of Python dicts.
+
+        Parameters
+        ----------
+        pvp_data: list of Python dicts
+            List of Python dicts (one for each channel) mapping parameter names
+            to NumPy arrays of data. See PVPBlock.toListOfDicts() for more info
+            on the structure expected here
+        metadata: cphd.Metadata
+            Metadata used to create this PVPBlock
+        """
+        pvp_block = PVPBlock(metadata.pvp, metadata.data)
+
+        # Methods to check if an optional PVP param is set
+        # Using .lower() since name capitalization is not 100% consistent
+        has_optional_param_methods = {name.lower(): getattr(pvp_block, name)
+                                      for name in dir(pvp_block)
+                                      if name.startswith('has')}
+
+        pvp_info = {}
+
+        # Set required and optional PVP parameters
+        for name in dir(metadata.pvp):
```

Comment:
There's a bit of code that's duplicated here from above. Here, we need to get the list of `PVPBlock` setter methods for all default PVP params (required ones plus any optional parameters that are in use) and ideally build a dict that maps PVP param names to setter methods